File: server/java/src/main/java/com/onlinepayments/example/api/validators/common/mandate/MandateValidator.java
package com.onlinepayments.example.api.validators.common.mandate;

import com.onlinepayments.example.api.exceptions.ApiValidationException;
import com.onlinepayments.example.api.validators.common.address.AdressValidator;
import com.onlinepayments.example.api.validators.common.iban.IbanValidator;
import com.onlinepayments.example.service.domain.common.enums.Country;
import com.onlinepayments.example.service.domain.payments.Mandate;

import static com.onlinepayments.example.api.validators.BaseValidator.isNullOrEmpty;
import static com.onlinepayments.example.api.validators.BaseValidator.isValidUrl;

public class MandateValidator {
    public static void validateMandate(Mandate mandate) {
        if (mandate == null) {
            throw new ApiValidationException("Mandate is required for DIRECT_DEBIT payments.");
        }

        if (isNullOrEmpty(mandate.getCustomerReference())) {
            throw new ApiValidationException("The CustomerReference field is required.");
        }

        if (mandate.getRecurrenceType() == null) {
            throw new ApiValidationException("The RecurrenceType field is required.");
        }

        if (mandate.getSignatureType() == null) {
            throw new ApiValidationException("The SignatureType field is required.");
        }

        if (mandate.getCustomerReference().length() > 35) {
            throw new ApiValidationException("The CustomerReference field must be shorter than 36 characters.");
        }

        if (isNullOrEmpty(mandate.getMandateReference()) && isNullOrEmpty(mandate.getIban())) {
            throw new ApiValidationException(
                    "The IBAN field is required when mandate reference is not provided."
            );
        }

        if (!isNullOrEmpty(mandate.getIban())) {

            if (mandate.getIban().length() > 50) {
                throw new ApiValidationException(
                        "The IBAN field must be less than 51 characters."
                );
            }

            Country country = null;
            if (mandate.getAddress() != null) {
                country = mandate.getAddress().country();
            }

            IbanValidator.validateIban(mandate.getIban(), country);
        }

        if (!isNullOrEmpty(mandate.getReturnUrl()) && !isValidUrl(mandate.getReturnUrl())) {
            throw new ApiValidationException("The ReturnUrl field is invalid.");
        }

        if (isNullOrEmpty(mandate.getMandateReference())) {
            AdressValidator.validateAddress(mandate.getAddress());
        }

        if (isNullOrEmpty(mandate.getMandateReference()) && mandate.getAddress() == null) {
            throw new ApiValidationException("Address is required when mandate reference is not provided.");
        }
    }
}

File: server/python/app/presentation/validators/common/mandate/mandate_validator.py
from urllib.parse import urlparse
from app.application.domain.payments.mandate import Mandate
from app.application.domain.enums.payments.recurrence_type import RecurrenceType
from app.application.domain.enums.payments.signature_type import SignatureType
from app.presentation.validators.common.address.address_validator import AddressValidator
from app.presentation.validators.common.iban.iban_validator import IbanValidator

class MandateValidator:

    def validate(self, mandate: Mandate, prefix: str = "Mandate") -> list[str]:
        errors = []

        if not mandate:
            errors.append(f"The {prefix} field is required.")
            return errors

        customer_reference = (mandate.customer_reference or "").strip()
        mandate_reference = (mandate.mandate_reference or "").strip()
        iban = (mandate.iban or "").strip()
        recurrence_type = mandate.recurrence_type
        signature_type = mandate.signature_type
        return_url = (mandate.return_url or "").strip()
        address = mandate.address

        if not customer_reference:
            errors.append("The CustomerReference field is required.")
        elif len(customer_reference) > 35:
            errors.append("The CustomerReference field must be shorter than 36 characters.")

        if not mandate_reference and not iban:
            errors.append("IBAN is required when mandate reference is not provided.")

        if not recurrence_type:
            errors.append("The RecurrenceType field is required.")
        elif not RecurrenceType.has_value(recurrence_type.value if hasattr(recurrence_type, 'value') else recurrence_type):
            errors.append("The RecurrenceType field is invalid.")

        if not signature_type:
            errors.append("The SignatureType field is required.")
        elif not SignatureType.has_value(signature_type.value if hasattr(signature_type, 'value') else signature_type):
            errors.append("The SignatureType field is invalid.")

        if return_url:
            try:
                parsed = urlparse(return_url)
                if parsed.scheme not in ("http", "https") or not parsed.netloc:
                    errors.append("The ReturnUrl field is invalid.")
                _ = parsed.port
            except ValueError:
                errors.append("The ReturnUrl field is invalid.")
            except Exception:
                errors.append("The ReturnUrl field is invalid.")

        if not mandate_reference:
            if not address:
                errors.append("Address is required when mandate reference is not provided.")
            else:
                errors.extend(AddressValidator.validate(address, "Mandate.Address"))

        if iban and not mandate_reference:
            if len(iban) > 50:
                errors.append("The IBAN field must be shorter than 51 characters.")
            else:
                country = address.country if address else None
                iban_errors = IbanValidator().validate(iban, country)
                errors.extend(iban_errors)

        return errors